xc/device.ml

```diff
@@ -2672,21 +2672,35 @@ module Backend = struct
              Unix.close save_fd)
 
 
-      (* Wait for QEMU's event socket to appear. *)
+      (* Wait for QEMU's event socket to appear. Connect to it
+       * to make sure it is ready. *)
       let wait_event_socket ~task ~name ~domid ~timeout =
         let finished = ref false in
         let timeout_ns = Int64.of_float (timeout *. Mtime.s_to_ns) in
+        let now = Mtime_clock.now () in
         let target =
-          match Mtime.add_span (Mtime_clock.now ()) (Mtime.Span.of_uint64_ns timeout_ns) with
+          match Mtime.(add_span now (Span.of_uint64_ns timeout_ns)) with
           | None -> raise (Ioemu_failed (name, "Timeout overflow"))
           | Some x -> x in
-        while Mtime.is_earlier (Mtime_clock.now ()) ~than:target && not !finished do
-          Xenops_task.check_cancelling task;
-          if Sys.file_exists (qmp_event_path domid) then
-            finished := true
-          else
-            ignore (Unix.select [] [] [] 0.05)
-        done;
+        let path   = qmp_event_path domid in
+        let socket = Unix.(socket PF_UNIX SOCK_STREAM 0) in
+        finally (* make sure we don't leak socket *)
+          (fun () ->
+             while Mtime.is_earlier (Mtime_clock.now ()) ~than:target
+                   && not !finished do
+               Xenops_task.check_cancelling task; (* might raise exn *)
+               if Sys.file_exists path then
+                 try
+                   Unix.connect socket Unix.(ADDR_UNIX path);
+                   finished := true
+                 with e ->
+                   debug "QMP event socket for domid %d not yet ready: %s"
+                     domid (Printexc.to_string e);
+                   Unix.sleepf 0.1
```

Comment:
I was thinking we could use the same timeout here and unify the branches, but I don't see a nice way of doing that (perhaps by moving the connect check to a function of its own, probably not worth it).

Reply:
I believe that `Unix.sleep(f)` blocks the entire process, and `Thread.delay` should be used instead.

Reply:
Notice that the old code used `select` for sleeping. I'm changing it to use `Thread.delay` which should be safe in any case.

Reply:
Looks like they are aliases, but Thread.delay more clearly expresses the intent I think: https://github.com/ocaml/ocaml/blob/c244e2db6e6a0a227c91a615d0d525c942126ede/Changes#L845-L846

Reply:
Looks like it is fine now, but in the past it used `sleep(3)`, which is not thread safe.

Reply:
At least on Linux, `sleep(3)` is just a wrapper around `nanosleep()` which is thread safe.